test/Spec.hs

```diff
@@ -0,0 +1,357 @@
+size_post_condition key priority val depq =
+    property $ D.size (D.insert key priority val depq) >= D.size depq
+
+topK_post_condition :: Int -> TestDEPQ -> Property
+topK_post_condition num depq =
+    let s = D.topK num depq
+    in if num <= (D.size depq) && num > 0
+        then Seq.length s === num
+        else Seq.length s === 0
+
+bottomK_post_condition :: Int -> TestDEPQ -> Property
+bottomK_post_condition num depq =
+    let s = D.bottomK num depq
+    in if num <= (D.size depq) && num > 0
+        then Seq.length s === num
+        else Seq.length s === 0
+
+popMin_post_condition :: TestDEPQ -> Property
+popMin_post_condition depq = 
+    let mPoped = D.popMin depq
+    in if D.size depq == 0
+        then mPoped === Nothing
+        else property $ maybe
+          False
+          (\(_val, depq') -> D.size depq' < D.size depq)
+          mPoped
+
+popMax_post_condition :: TestDEPQ -> Property
+popMax_post_condition depq = 
+    let mPoped = D.popMax depq
+    in if D.size depq == 0
+        then mPoped === Nothing
+        else property $ maybe
+          False
+          (\(_val, depq') -> D.size depq' < D.size depq)
+          mPoped
+
+--------------------------------------------------------------------------------
+-- Metamorphic
+-- Instead of testing single function, we're now going to test multiple set of
+-- functions
+--------------------------------------------------------------------------------
+
+metamorphicSpec :: Spec
+metamorphicSpec = describe "Metamorphic" $ do
+    prop "insert twice, findMax" insert_insert_findMax
+    prop "insert twice, findMin" insert_insert_findMin
+
+-- Insert 2 key value pairs, test that findMax return expected value
+insert_insert_findMax :: (Int, Priority, Val) -> (Int, Priority, Val) -> Property
+insert_insert_findMax t1@(k1, p1, v1) t2@(k2, p2, v2)
+      | k1 == k2 = D.findMax depq === Just t2
+      | p1 < p2 = D.findMax depq === Just t2
+      | p1 > p2 = D.findMax depq === Just t1
+      | k1 > k2 = D.findMax depq === Just t2
+      | otherwise = D.findMax depq === Just t1
+  where
+    depq = D.insert k2 p2 v2 $ D.insert k1 p1 v1 D.empty
+
+-- Insert 2 key value pairs, test that findMin return expected value
+insert_insert_findMin :: (Int, Priority, Val) -> (Int, Priority, Val) -> Property
+insert_insert_findMin t1@(k1, p1, v1) t2@(k2, p2, v2)
+      | k1 == k2 = D.findMin depq === Just t2
+      | p1 < p2 = D.findMin depq === Just t1
+      | p1 > p2 = D.findMin depq === Just t2
+      | k1 > k2 = D.findMin depq === Just t2
+      | otherwise = D.findMin depq === Just t1
+  where
+    depq = D.insert k2 p2 v2 $ D.insert k1 p1 v1 D.empty
+
+--------------------------------------------------------------------------------
+-- Model based testing
+-- Here, we test against a data structure that would act as an model
+-- (i.e. Something that would behave similar to DEPQ and is well tested)
+--------------------------------------------------------------------------------
```

Comment:
Model based testing might be a bit of over-engineering. 

1. It's hard to understand the concept of it.
2. It's re-implementing set of functions.
3. Maintaining them might become difficult.

If you are willing to add test on other categories (e.g. post-condition, metamorphic) you might as well drop this one.